services/fluxmonitor.go

```diff
@@ -0,0 +1,272 @@
+	"net/url"
+	"sort"
+	"time"
+
+	"github.com/shopspring/decimal"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/tidwall/gjson"
+)
+
+type FluxMonitorConfig struct {
+	Adapters          []url.URL
+	From              string
+	To                string
+	Multiply          int64
+	Threshold         decimal.Decimal
+	AbsoluteThreshold decimal.Decimal
+	Heartbeat         time.Duration
+	PollInterval      time.Duration
+}
+
+func ParseFMSpec(jsonSpec json.RawMessage) FluxMonitorConfig {
+	var fmConfig FluxMonitorConfig
+	res := gjson.GetBytes(jsonSpec, "feeds.#.url")
+	var adapters []url.URL
+	for _, adapter := range res.Array() {
+		url, _ := url.Parse(adapter.String())
+		adapters = append(adapters, *url)
+	}
+	fmConfig.Adapters = adapters
+	fmConfig.From = gjson.GetBytes(jsonSpec, "requestData.data.from").String()
+	fmConfig.To = gjson.GetBytes(jsonSpec, "requestData.data.to").String()
+	fmConfig.Multiply = gjson.GetBytes(jsonSpec, "precision").Int()
+	fmConfig.Threshold, _ = decimal.NewFromString(gjson.GetBytes(jsonSpec, "threshold").String())
+	fmConfig.AbsoluteThreshold, _ = decimal.NewFromString(gjson.GetBytes(jsonSpec, "absoluteThreshold").String())
+	fmConfig.Heartbeat, _ = time.ParseDuration(gjson.GetBytes(jsonSpec, "idleTimer.duration").String())
+	fmConfig.PollInterval, _ = time.ParseDuration(gjson.GetBytes(jsonSpec, "pollTimer.period").String())
+	return fmConfig
+}
+
+type FluxAggregatorState struct {
+	currentAnswer decimal.Decimal
+}
+
+type AdapterResponse struct {
+	Price decimal.Decimal `json:"result"`
+	// might need error response as well
+}
+type FluxMonitor struct {
+	state  FluxAggregatorState
+	config FluxMonitorConfig
+
+	// subscriber subscriber.ISubscriber
+	latestResult decimal.Decimal
+
+	// quitOnce     sync.Once
+
+	// chBlockchainEvents chan subscriber.Event
+	chDeviation chan decimal.Decimal
+	// chNewround  chan FluxAggregatorState
+	// chClose chan struct{}
+}
+
+func NewFluxMonitor(config FluxMonitorConfig, triggerJobRun chan decimal.Decimal) *FluxMonitor {
+	logger.Info(fmt.Sprintf("New FluxMonitor with config: %+v", config))
+	srv := FluxMonitor{
+		config: config,
+	}
+	go srv.startPoller()
+	srv.hitTrigger(triggerJobRun)
+	return &srv
+}
+
+func (fm *FluxMonitor) hitTrigger(triggerJobRun chan decimal.Decimal) {
+	fm.chDeviation = make(chan decimal.Decimal)
+	ticker := time.NewTicker(fm.config.Heartbeat)
+	defer ticker.Stop()
+	for {
+
+		select {
+		// case <-fm.chNewround:
+		// 	fmt.Println("New round started")
+		// 	fm.state.currentAnswer = fm.latestResult
+		// 	fmt.Println("New answer: ", fm.state.currentAnswer)
+		case <-fm.chDeviation:
+			logger.Info("Deviation threshold hit")
+			fm.state.currentAnswer = <-fm.chDeviation
+			triggerJobRun <- fm.state.currentAnswer
+			logger.Info("New answer: ", fm.state.currentAnswer)
+		case <-ticker.C:
+			logger.Info("Heartbeat")
+			// if adapters not working this is going to resubmit an old value
+			fm.state.currentAnswer = fm.latestResult
```

Comment:
We can't update the state's `currentAnswer` until we actually get an event that it's been updated.

Reply:
Same goes for L94 too